filebeat/spooler/spooler.go

```diff
@@ -61,23 +59,32 @@ func (s *Spooler) Start() {
 func (s *Spooler) run() {
 	defer s.wg.Done()
 
-	ticker := time.NewTicker(s.config.idleTimeout / 2)
-
 	logp.Info("Starting spooler: spool_size: %v; idle_timeout: %s",
 		s.config.spoolSize, s.config.idleTimeout)
 
+	timer := time.NewTimer(s.config.idleTimeout)
+
 loop:
 	for {
 		select {
 		case <-s.exit:
-			ticker.Stop()
+			timer.Stop()
 			break loop
 		case event := <-s.Channel:
 			if event != nil {
-				s.queue(event)
+				flushed := s.queue(event)
+				if flushed {
+					// Stop timer and drain channel. See https://golang.org/pkg/time/#Timer.Reset
+					if !timer.Stop() {
+						<-timer.C
+					}
+					timer.Reset(s.config.idleTimeout)
+				}
 			}
-		case <-ticker.C:
-			s.timedFlush()
+		case <-timer.C:
+			debugf("Flushing spooler because of timeout. Events flushed: %v", len(s.spool))
+			s.flush()
+			timer.Reset(s.config.idleTimeout)
```

Comment:
That means we would stop and drain the timer before the flush and reset it only afterwards?